draft-ietf-webpush-encryption.md

```diff
@@ -108,170 +101,277 @@ encryption scheme by applications that use push messaging.  An agent that only
-[I-D.ietf-httpbis-encryption-encoding].  This authentication secret (see
-{{auth}}) is generated by the user agent and shared with the application server.
+## Key Derivation Context {#context}
 
+The derivation of the content encryption key and nonce uses an additional
+context string.
+
+The context is comprised of a label of "P-256" encoded in ASCII (that is, the
+octet sequence 0x50, 0x2d, 0x32, 0x35, 0x36), a zero-valued octet, the length of
+the User Agent public key (65 octets) encoded as a two octet unsigned integer in
+network byte order, the User Agent public key, the length of the Application
+Server public key (65 octets), and the Application Server public key.
+
+~~~ inline
+   context = label || 0x00 ||
+               length(ua_public) || ua_public ||
+               length(as_public) || as_public
+~~~
 
-## Push Message Content Encryption {#c-e}
+## Encryption Summary {#summary}
+
+This results in a the final content encryption key and nonce generation using
+the following sequence, which is shown here in pseudocode with HKDF expanded
+into separate discrete steps using HMAC with SHA-256:
+
+~~~ inline
+   -- For a User Agent:
+   ecdh_secret = ECDH(ua_private, as_public)
+   auth_secret = random(16)
+
+   -- For an Application Server:
+   ecdh_secret = ECDH(as_private, ua_public)
+   auth_secret = <from User Agent>
+
+   -- For both:
+   auth_info = "Content-Encoding: auth" || 0x00
+   PRK_combine = HMAC-SHA-256(auth_secret, ecdh_secret)
+   IKM = HMAC-SHA-256(PRK_combine, auth_info || 0x01)
+   context = "P-256" || 0x00 ||
+             0x00 || 0x41 || ua_public ||
+             0x00 || 0x41 || as_public
+   salt = random(16)
+   PRK = HMAC-SHA-256(salt, IKM)
+   cek_info = "Content-Encoding: aesgcm" || 0x00 || context
+   CEK = HMAC-SHA-256(PRK, cek_info || 0x01)[0..15]
+   nonce_info = "Content-Encoding: nonce" || 0x00 || context
+   NONCE = HMAC-SHA-256(PRK, nonce_info || 0x01)[0..11]
+~~~
 
-The Application Server then encrypts the payload.  Header fields are populated
-with base64url encoded [RFC7515] values:
+Note that this omits the exclusive OR of the final nonce with the record
+sequence number, since implementations of this specification are not required to
+support multiple records (see {{restrict}}) and the sequence number of the first
+record is zero.
 
-* the salt is added to the `salt` parameter of the Encryption header field; and
 
-* the public key for its DH or ECDH key pair is placed in the `dh` parameter of
-  the Crypto-Key header field.
+# Restrictions on Use of "aesgcm" Content Coding {#restrict}
 
-An application server MUST encrypt a push message with a single record.  This
+An Application Server MUST encrypt a push message with a single record.  This
 allows for a minimal receiver implementation that handles a single record.  If
 the message is 4096 octets or longer, the `rs` parameter MUST be set to a value
 that is longer than the encrypted push message length.
```

Comment:
I'd be fine with defining the `rs` to default to `record_size + 1` as opposed to `4096`, as done by httpbis-encryption-encoding. Since that specification defines a default, we do need language addressing this.